File: llm-smartphones-bench/executorch-app/app/src/main/java/com/example/executorchllamademo/DeviceInfoUtils.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.example.executorchllamademo;

import android.app.ActivityManager;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.BatteryManager;
import android.os.Build;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStreamReader;

/**
 * Android-native equivalents of selected react-native-device-info APIs.
 */
public final class DeviceInfoUtils {
    private DeviceInfoUtils() {}

    /**
     * Returns total system memory in bytes.
     * Mirrors react-native-device-info getTotalMemory.
     */
    public static long getTotalMemory(Context context) {
        try {
            ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
            ActivityManager.MemoryInfo mi = new ActivityManager.MemoryInfo();
            am.getMemoryInfo(mi);
            return mi.totalMem > 0L ? mi.totalMem : readMemTotalFromProc();
        } catch (Throwable e) {
            return readMemTotalFromProc();
        }
    }

    /**
     * Returns battery level as a value in [0.0, 1.0].
     * Mirrors react-native-device-info getBatteryLevel.
     */
    public static float getBatteryLevel(Context context) {
        try {
            IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
            Intent batteryStatus = context.registerReceiver(null, filter);
            int level = batteryStatus != null ? batteryStatus.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) : -1;
            int scale = batteryStatus != null ? batteryStatus.getIntExtra(BatteryManager.EXTRA_SCALE, -1) : -1;
            return (level >= 0 && scale > 0) ? (float) level / (float) scale : 0f;
        } catch (Throwable e) {
            return 0f;
        }
    }

    /** Model name of the device, e.g., "Pixel 7 Pro". */
    public static String getModel() {
        return Build.MODEL != null ? Build.MODEL : "";
    }

    /** Device brand, e.g., "google", "samsung". */
    public static String getBrand() {
        return Build.BRAND != null ? Build.BRAND : "";
    }

    /** Operating system name. */
    public static String getSystemName() {
        return "Android";
    }

    /** Operating system version string, e.g., "14" or "8.1.0". */
    public static String getSystemVersion() {
        return Build.VERSION.RELEASE != null ? Build.VERSION.RELEASE : String.valueOf(Build.VERSION.SDK_INT);
    }

    private static long readMemTotalFromProc() {
        try {
            File file = new File("/proc/meminfo");
            if (!file.exists() || !file.canRead()) return 0L;
            try (FileInputStream fis = new FileInputStream(file);
                 BufferedReader br = new BufferedReader(new InputStreamReader(fis))) {
                String firstLine = br.readLine();
                if (firstLine == null) return 0L;
                // Expected format: "MemTotal:        12345678 kB"
                String[] parts = firstLine.trim().split("\\s+");
                if (parts.length >= 2) {
                    try {
                        long kb = Long.parseLong(parts[1]);
                        return kb * 1024L;
                    } catch (NumberFormatException e) {
                        return 0L;
                    }
                } else {
                    return 0L;
                }
            }
        } catch (Throwable e) {
            return 0L;
        }
    }
}

File: test/android/MLCChat/app/src/main/java/ai/mlc/mlcchat/DeviceInfoUtils.kt
package ai.mlc.mlcchat

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.InputStreamReader

object DeviceInfoUtils {

    fun getTotalMemory(context: Context): Long {
        return try {
            val am = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            val mi = ActivityManager.MemoryInfo()
            am.getMemoryInfo(mi)
            if (mi.totalMem > 0L) mi.totalMem else readMemTotalFromProc()
        } catch (_: Throwable) {
            readMemTotalFromProc()
        }
    }

    fun getBatteryLevel(context: Context): Float {
        return try {
            val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
            val batteryStatus: Intent? = context.registerReceiver(null, filter)
            val level = batteryStatus?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
            val scale = batteryStatus?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
            if (level >= 0 && scale > 0) level.toFloat() / scale.toFloat() else 0f
        } catch (_: Throwable) {
            0f
        }
    }

    fun getModel(): String = Build.MODEL ?: ""

    fun getBrand(): String = Build.BRAND ?: ""

    fun getSystemName(): String = "Android"

    fun getSystemVersion(): String = Build.VERSION.RELEASE ?: Build.VERSION.SDK_INT.toString()

    private fun readMemTotalFromProc(): Long {
        return try {
            val file = File("/proc/meminfo")
            if (!file.exists() || !file.canRead()) return 0L
            FileInputStream(file).use { fis ->
                BufferedReader(InputStreamReader(fis)).use { br ->
                    val firstLine = br.readLine() ?: return 0L
                    val parts = firstLine.trim().split(Regex("\\s+"))
                    if (parts.size >= 2) {
                        val kb = parts[1].toLongOrNull() ?: return 0L
                        kb * 1024L
                    } else 0L
                }
            }
        } catch (_: Throwable) {
            0L
        }
    }
}


